src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -344,13 +304,143 @@ let verify_transitions_and_build_breadcrumbs ~logger
+          in
+          Deferred.Or_error.return (acc, initial_state_hash) )
+  in
+  let trees_of_transitions =
+    Option.fold
+      (Non_empty_list.of_list_opt transitions_with_initial_validation)
+      ~init:subtrees ~f:(fun _ transitions ->
+        [Rose_tree.of_non_empty_list ~subtrees transitions] )
+  in
+  let open Deferred.Let_syntax in
+  match%bind
+    Transition_handler.Breadcrumb_builder.build_subtrees_of_breadcrumbs ~logger
+      ~precomputed_values ~verifier ~trust_system ~frontier ~initial_hash
+      trees_of_transitions
+  with
+  | Ok result ->
+      Deferred.Or_error.return (transitions_with_initial_validation, result)
+  | Error e ->
+      List.map transitions_with_initial_validation
+        ~f:Cached.invalidate_with_failure
+      |> ignore ;
+      Deferred.Or_error.fail e
+
+let download_transitions_in_chunks ~logger ~trust_system ~network ~num_peers
+    ~preferred_peer ~maximum_download_size ~hashes_of_missing_transitions =
+  let%bind random_peers = Coda_networking.random_peers network num_peers in
+  Deferred.Or_error.List.map
+    (partition maximum_download_size hashes_of_missing_transitions)
+    ~how:`Parallel ~f:(fun hashes ->
+      Deferred.Or_error.find_map_ok (preferred_peer :: random_peers)
+        ~f:(fun peer ->
+          let open Deferred.Or_error.Let_syntax in
+          let%bind transitions =
+            Coda_networking.get_transition_chain network peer hashes
+          in
+          Coda_metrics.(
+            Gauge.set
+              Transition_frontier_controller
+              .transitions_downloaded_from_catchup
+              (Float.of_int (List.length transitions))) ;
+          if not @@ verify_against_hashes transitions hashes then (
+            let error_msg =
+              sprintf
+                !"Peer %{sexp:Network_peer.Peer.t} returned a list that is \
+                  different from the one that is requested."
+                peer
+            in
+            Trust_system.(
+              record trust_system logger peer.host
+                Actions.(Violated_protocol, Some (error_msg, [])))
+            |> don't_wait_for ;
+            Deferred.Or_error.error_string error_msg )
+          else
+            Deferred.Or_error.return
+            @@ List.map2_exn hashes transitions ~f:(fun hash transition ->
+                   let transition_with_hash =
+                     With_hash.of_data transition ~hash_data:(Fn.const hash)
+                   in
+                   Envelope.Incoming.wrap ~data:transition_with_hash
+                     ~sender:(Envelope.Sender.Remote (peer.host, peer.peer_id))
+               ) ) )
+
+exception Breadcrumb_Error of Error.t
+
+let verify_transitions_and_build_breadcrumbs_in_chunks ~logger
+    ~precomputed_values ~trust_system ~verifier ~frontier
+    ~unprocessed_transition_cache ~transitions_chunks ~target_hash ~subtrees =
+  try
+    let previous_transitions_with_initial_validation = ref [] in
```

Comment:
is it possible to avoid a reference, maybe by using `List.foldi` and accumulating this list?

Reply:
Yeah! Jiawei also mentioned this. Changes are in progress. Thanks!